src/Microsoft.OData.CodeGen/Templates/ODataT4CodeGenerator.cs

```diff
@@ -2908,6 +2919,166 @@ internal void SetPropertyIdentifierMappingsIfNameConflicts(string typeName, IEdm
+                }
+            }
+            else
+            {
+                IdentifierMappings.Add(property.Name, validName);
+            }
+        }
+    }
+
+    /// <summary>
+    /// Regex match any non-identifier characters, used to generate valid identifiers<br />
+    /// See C# definitions: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#643-identifiers <br />
+    /// See VB definitions: https://learn.microsoft.com/en-us/dotnet/visual-basic/reference/language-specification/lexical-grammar#identifiers <br />
+    /// See Unicode character class definitions: https://www.unicode.org/reports/tr18/#General_Category_Property
+    /// </summary>
+    private static readonly Regex RegexMatchNonIdentifier = new Regex(@"[^\p{L}\p{Nl}\p{Mn}\p{Mc}\p{Nd}\p{Pc}\p{Cf}]", RegexOptions.Compiled);
+
+    /// <summary>
+    /// Default prefix character used to prefix non-alpha characters in identifiers
+    /// </summary>
+    private const string DefaultPrefixCharacter = "_";
+
+    /// <summary>
+    /// Get a valid C#/VB identifier for <paramref name="name"/> based on C#/VB identifier specification<br />
+    /// - Removes characters not permitted in C#/VB identifiers<br />
+    /// - Prefixes <paramref name="name"/>with underscore (_) if it doesn't start with a letter
+    /// </summary>
+    /// <remarks>
+    /// C# and VB use very similar rules for identifier names<br />
+    /// See C# definitions: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#643-identifiers <br />
+    /// See VB definitions: https://learn.microsoft.com/en-us/dotnet/visual-basic/reference/language-specification/lexical-grammar#identifiers
+    /// </remarks>
+    /// <param name="name">Identifier to validate and modify</param>
+    /// <returns>Modified <paramref name="name"/> or <paramref name="name"/> unchanged if no changes are needed</returns>
+    private string GetValidIdentifier(string name)
+    {
+        var codeDomProvider = (this.context.TargetLanguage == LanguageOption.CSharp ? CSharpProvider : VBProvider);
+        if (codeDomProvider.IsValidIdentifier(name))
+        {
+            return name;
+        }
+
+        var segments = RegexMatchNonIdentifier.Split(name);
+        var validName = new StringBuilder();
+        var isFirst = true;
+        foreach (var segment in segments.Where(token => !string.IsNullOrWhiteSpace(token)))
+        {
+            var titleCaseSegment = Customization.CustomizeNaming(segment);
+
+            if (isFirst && !this.context.EnableNamingAlias)
+            {
+                validName.Append(segment);
+            }
+            else
+            {
+                validName.Append(titleCaseSegment);
+            }
+
+            isFirst = false;
+        }
+
+        if (!char.IsLetter(validName.ToString(), 0) && !validName.ToString().StartsWith(DefaultPrefixCharacter))
```

Comment:
Indexing the string builder directly would be far more efficient than doing `ToString()` just to read the first char. We're also calling `ToString()` multiple times which feels unnecessarily wasteful.

```suggestion
        if (!char.IsLetter(validName[0]) && validName[0] != DefaultPrefixCharacter)
```

![image](https://github.com/OData/ODataConnectedService/assets/8460169/9e8aad41-0f4d-4cc1-bd97-3e28e3f5d2a4)
Benchmark code: https://gist.github.com/habbes/f244a5991f56196f2c3030ecb2a7f8fb